libgadget/blackhole.c

```diff
@@ -766,8 +770,17 @@ blackhole_feedback_ngbiter(TreeWalkQueryBHFeedback * I,
         O->BH_CountProgs += BHP(other).CountProgs;
         /* Leave the swallowed BH mass around
          * so we can work out mass at merger. */
-        O->Mass += (P[other].Mass);
+        
         O->BH_Mass += (BHP(other).Mass);
+        
+        /* Prevent large DynMass accumulates during merger */
```

Comment:
This seems odd to me?

Reply:
If (I->BH_Mass > SeedBHDynMass && BHP(other).Mass > SeedBHDynMass ) {
   /* both have grown beyond Seed, directly add the masses*/
   I->BH_Mass  += BHP(other)->Mass;
   I->Mass += P[other]->mass
}
If (I->BH_Mass > SeedBHDynMass && BHP(other).Mass < SeedBHDynMass ) {
   I->BH_Mass  += BHP(other)->Mass
}
If (I->BH_Mass < SeedBHDynMass && BHP(other).Mass > SeedBHDynMass ) {
   I->BH_Mass  += BHP(other)->Mass;
   I->Mass = P[other]->mass
}
If (I->BH_Mass < SeedBHDynMass && BHP(other).Mass < SeedBHDynMass ) {
   I->BH_Mass  += BHP(other)->Mass;
   I->Mass = SeedBHDynMass
}

Reply:
I think that the SeedBHDynMass should be almost totally decoupled from the BH_Mass. I also think that merging a seed mass black hole should do *something* just not as much as a full BHDynMass. So I propose:

O->BH_Mass += BHP(other)->Mass;
/* This code ensures that the extra dynamic mass created during seeding does not persist after mergers*/
If (I->Mass > SeedBHDynMass && P[other].Mass > SeedBHDynMass ) {
/* both have grown beyond Seed, directly add the masses*/
O->Mass += P[other]->mass
}
If (I->Mass <= SeedBHDynMass && P[other].Mass >= SeedBHDynMass ) {
O->Mass = P[other]->mass += (one gas particle mass);
}
If (I->Mass <= SeedBHDynMass && P[other].Mass <= SeedBHDynMass ) {
O->Mass = SeedBHDynMass 
/* Except this is not right: we have 'lost' the mass of a gas particle if a BH was seeded just before accretion*/
}

Reply:
Note that O->Mass is the 'BH_accreted_Mass' or accmass instead of the final dynamic mass (see feedback_reduce, feedback_postprocess)

You actually mean 
If (I->Mass <= SeedBHDynMass && P[other].Mass <= SeedBHDynMass ) {O->Mass += 0}
Is that right?

But that will cause, e.g.,  {Mdyn: 1e7, Mint:9e6} + {Mdyn:1e7, Mint:9e6} = {Mdyn:1e7, Mint:1.8e7} and it will quickly swallow 4 gas particles after merger.  Not sure whether it will cause issue.

Reply:
My previous implementation enforces {Mdyn: 1e7, Mint:9e6} + {Mdyn:1e7, Mint:9e6} = {Mdyn:1.8e7, Mint:1.8e7}, but that will fail to swallow some gas particle for the good of mass conservation. 
We might need to modify the swallow of gas carefully anyway if we want to improve mass conservation by swallowing gas.